Source/WebKit/WebProcess/WebPage/Cocoa/UnifiedTextReplacementController.mm

```diff
@@ -669,45 +542,126 @@ static void replaceContentsInRange(WebCore::Document& document, const WebCore::S
     if (m_contextRanges.isEmpty())
         return;
 
-    if (!m_webPage) {
+    RefPtr document = this->document();
+    if (!document) {
         ASSERT_NOT_REACHED();
         return;
     }
 
+    auto selectionRange = document->selection().selection().firstRange();
+    if (!selectionRange)
+        return;
+
+    if (!document->selection().isCaret())
+        return;
+
+    auto nodeAndMarker = findReplacementMarkerContainingRange(*document, *selectionRange);
+    if (!nodeAndMarker)
+        return;
+
+    auto& [node, marker] = *nodeAndMarker;
+    auto data = std::get<WebCore::DocumentMarker::UnifiedTextReplacementData>(marker.data());
+
+    m_webPage->textReplacementSessionUpdateStateForReplacementWithUUID(data.sessionUUID, WebTextReplacementData::State::Active, data.uuid);
+}
+
+// MARK: Private instance helper methods.
+
+std::optional<WebCore::SimpleRange> UnifiedTextReplacementController::contextRangeForSessionWithUUID(const WTF::UUID& uuid) const
+{
+    RefPtr liveRange = m_contextRanges.get(uuid);
+    return WebCore::makeSimpleRange(liveRange);
```

Comment:
Minor nit - I'm not sure this temporary variable adds a lot — maybe just do `return WebCore::makeSimpleRange(m_contextRanges.get(uuid));`?